Adjust API 'editoffer' to PR 5651 (include extraData field when editing offer)

ghubstan opened the pull request:
PR https://github.com/bisq-network/bisq/pull/5651 ensures payment acct changes are saved when the UI edits an offer.

Although the API `editoffer` command does not yet support editing payment account details, this series of PRs must adjust to 5651 (merged today) to avoid file conflicts.

This PR is based on https://github.com/bisq-network/bisq/pull/5663, which should be reviewed and merged before this one.

ghubstan committed:
Refactor EditOfferDataModel for new editoffer api method

- Define set of editable offer payload fields in MutableOfferPayloadFields.

- Move bulk of EditOfferDataModel#onPublishOffer logic to OfferUtil.

ghubstan committed:
Fix comment

ghubstan committed:
Add OfferInfo field isActivated, rpc EditOffer to proto

ghubstan committed:
Add server/core editOffer, adjust getMyOffer(s) impls

- Add editOffer to GrpcOffersService, CoreApi, CoreOffersService.

- Set editOffer call rate meter to 1 / minute.

- Use new EditOfferValidator to verify editOffer params OK.

- Adust getMyOffer(s) rpc impl and OfferInfo model to use OpenOffer
  for accessing activation state and trigger price.

ghubstan committed:
Refactor GrpcClient: request builders moved bisq.cli.request pkg

Reduces size of GrpcClient while allowing for additional bot-friendly
variations of the new grpc editOffer method.

ghubstan committed:
Add new EditOfferOptionParser and test

ghubstan committed:
Add editoffer method help

ghubstan committed:
Add editoffer to CLI

ghubstan committed:
Add editoffer api tests & minor apitest refactoring

ghubstan committed:
Remove chase quickpay acct test

ghubstan committed:
Remove unused import

ghubstan committed:
Fix problems found in codacy check

ghubstan committed:
Fix codacy issue

ghubstan committed:
Fix log arg spec bug

ghubstan committed:
Add bool isMyOffer to OfferInfo proto

ghubstan committed:
Fix editoffer validation bugs, tidy up CoreOffersService

ghubstan committed:
Show enable/trigger-price cols for 'getmyoffer'

ghubstan committed:
Improve 'editoffer' opt parsing, fix test pkg name

ghubstan committed:
Add editoffer test case, suppress annoying warnings

ghubstan committed:
Remove unused field

ghubstan committed:
Throw exception is edit altcoin offer is attempted

Support for editing BSQ offers is in place, but will be added
in another PR.

ghubstan committed:
Avoid duplicate test run

ghubstan committed:
Make codacy just a bit happier

ghubstan committed:
Force codacy check after codacy config change

ghubstan committed:
Include isMyOffer flag in API's trade/offer proto wrappers

Optionally displaying an ENABLED column in CLI side getoffer output
depends on the value of offer.isMyOffer, which is passed via new
boolean arguments to the trade & offer pojo builders.

ghubstan committed:
Pass isMyOffer flag to trade/offer proto wrappers from core services

ghubstan committed:
Add altcoin (bsq) offer editing validation check

BSQ offers are fixed-price only.  This change blocks an
attempt to change an altcoin offer to a margin price based
offer, or set a trigger price.

ghubstan committed:
Optionally show ENABLED column in CLI's getoffer(bsq) output

ghubstan committed:
Add BSQ offer editing tests to EditOfferTest

And log CLI's getoffer output to see getoffer formatting -- after adding
new ENABLED and TRIGGER-PRICE columns.

ghubstan committed:
Merge branch 'master' into 02-refactor-for-api-editoffer-method

ghubstan committed:
Merge branch '02-refactor-for-api-editoffer-method' into 03-api-editoffer-method

ghubstan committed:
Merge branch '03-api-editoffer-method' into 04-edit-bsq-offer

ghubstan committed:
Force rebuild after github action ECONNRESET

ghubstan committed:
Document api 'editoffer' usage

ghubstan committed:
Fix header

ghubstan committed:
Delete tmp main() method

ghubstan committed:
Rename and move private function

And make sure function is not duplicated CLI side logic.

ghubstan committed:
Do not duplicate Price.parse on CLI side for only one use case

ghubstan committed:
Remove old TODO because relevant refactoring was approved

ghubstan committed:
Fix typo 'enabled' -> 'enable'

ghubstan committed:
Always use Locale.US in CLI DecimalFormats

Avoid inconsistent CLI output decimal formats across different
systems' default locales.

ghubstan committed:
Merge branch 'master' into 05-update-api-beta-test-guide

ghubstan committed:
Do not start test harness deamons in dbg mode by default

ghubstan committed:
Add missing trigger-price param

ghubstan committed:
Fix peer add(offer) & remove(offer) event order problem

Use LinkedHashSet to maintain NetworkEnvelope ordering when Connection#onBundleOfEnvelopes
calls listeners.

Connection#onBundleOfEnvelopes builds a set from an ordered list of NetworkEnvelopes,
then calls listeners during set iteration.  The envelope list ordering is lost if a
HashSet is built, but maintained by switching to a LinkedHashSet.

Losing the envelope ordering becomes a problem if the peer receives a RemoveDataMessage
and an AddDataMessage in the same batch of envelopes, and relays them to listeners
in the wrong (random) order.  For example, an API 'editoffer' call may result in the
edited offer being added, then immediately remove from their UI's offer book.

ghubstan committed:
Revert "Fix peer add(offer) & remove(offer) event order problem"

This reverts commit add6536402daa36330711d6b73ad99157e00187e.

ghubstan committed:
Merge branch 'master' into 05-update-api-beta-test-guide

ghubstan committed:
Ensure UI OfferBook items are correctly added and removed

Using the API's CLI to edit offers can sometimes result in add/remove messages
being received on peers in the same batch of envolopes, and these messages
are sometimes passed to the UI in (1) add, (2) remove order.  This can result in
a newly edited offer being removed immediately after being added to the OfferBook
list. This change uses storage entry sequence number and storage entry payload
hash comparisons to avoid the problem.

- OfferBookListItem Added new constructor taking P2PDataStorage.ByteArray hashOfPayload,
  and int sequenceNumber params.  Added a new toString() method.

- OfferBook Added new checks on OfferBookListItem hashOfPayload and sequenceNumber while
  determining if offer candidates should be added or removed from the UI's OfferBook List.
  See OfferBook contructor's implementation of OfferBookChangedListener#onAdded and
  OfferBookChangedListener#onRemoved.  Added many comments explaining the add/remove rules,
  and plenty of debug statements to help trace the add/remove event process.

- OfferBookService#OfferBookChangedListener Added new P2PDataStorage.ByteArray hashOfPayload,
  and int sequenceNumber params to listener's onAdded and onRemoved method signatures.
  Added these two new paramater values to listener.onAdded and listener.onRemoved calls.

- TakeOfferDataModel Replaced unused, old tradeManager param in offerBook.removeOffer()
  with (null) P2PDataStorage.ByteArray hashOfPayload, and (-1) int sequenceNumber params.
  OfferBook will remove the candidate offer as before.

- MarketAlerts Adjusted onAdded() & onRemoved listener method signatures, even though
  new P2PDataStorage.ByteArray hashOfPayload, int sequenceNumber params are not used
  by the implementations.

ghubstan committed:
Revert logback debug config changes

(Accidentally included in last commit.)

ghubstan committed:
Merge branch 'master' into 06-safe-offerbook-add-remove-events

ghubstan committed:
Remove comment (question answered)

ghubstan committed:
Handle API's edit+disable offer use case in UI.

This change is a refactoring for handling the removal of a peer UI's offer
item when it is deactivated and edited in the same CLI `editoffer` command.

On the API side, an `editoffer --price=N --enable=false` command results
in the edited offer not being re-published.  On a UI peer's side, the edited
offer is not added to the peer's storage, and the peer's onRemoved(offer)
listener event does not find a storage entry with matching payload-hash.

This fix assumes an offer that is not in the local store should be removed
from the UI's view list -- when the onRemoved method's hashOfPayload does
not match the UI's view list item's hashOfPayload.

ghubstan committed:
Display Enabled=PENDING in CLI 'createoffer' output

A newly created offer has no OpenOffer+State (AVAILABLE || DEACTIVATED)
when displayed in the CLI's console.  This change adds a 'bool isMyPendingOffer'
to the OfferInfo proto + wrapper, and the CLI's console offer output formatter
uses it to determine if it should display a new offer's Enabled column value
as PENDING, instead of an ambiguous NO value.

ghubstan committed:
Merge branch 'master' into 08-handle-extradata-in-editoffer

Resolve file conflicts resulting from merging PRs
    https://github.com/bisq-network/bisq/pull/5577
    https://github.com/bisq-network/bisq/pull/5651

ghubstan committed:
Refactor 2 predicates as single predicate

Resolves https://github.com/bisq-network/bisq/pull/5659#discussion_r689631147

ghubstan committed:
Fix long method name

Resolves https://github.com/bisq-network/bisq/pull/5659#discussion_r687839333

ghubstan committed:
Merge branch '06-safe-offerbook-add-remove-events' into 07-show-new-offer-with-pending-status

ghubstan committed:
Merge branch '07-show-new-offer-with-pending-status' into 08-handle-extradata-in-editoffer

ghubstan committed:
Fix OfferBookService bug causing extra check in OfferBook.onRemoved

Hash of protectedStorageEntry (should be offerPayload) was sometimes
resulting in incorrect hash being sent to OfferBook listener methods
onAdded(offer,  hashOfPayload,  sequenceNumber), and
onRemoved(offer,  hashOfPayload,  sequenceNumber).
Hash of OfferPayload is correctly passed to listener with this change.

Sending the correct hash allows removal of a dubious code block that
removed a book view list item when hash compare failed, and no matching
offer existed in the OfferBookService.
See https://github.com/bisq-network/bisq/pull/5659#discussion_r689634240

ghubstan committed:
Remove unused methods

ghubstan committed:
Remove sequence-number checks from OfferBook

Checking offer payload hashes in OfferBook's onAdded and onRemove methods
is sufficient to prevent incorrect removal of offer list items from the
UI OfferBook view (where api 'editoffer' causes onRemoved to be called
after onAdded on peers).

sqrrm commented on desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:
```diff
@@ -83,44 +85,113 @@ public void onAdded(Offer offer) {
+        // onRemoved hashOfPayload parameter.  We may receive add/remove messages out of
+        // order (from api's 'editoffer'), and use the offer payload hash to
+        // ensure we do not remove an edited offer immediately after it was added.
+        if ((candidate.getHashOfPayload() == null || candidate.getHashOfPayload().equals(hashOfPayload))) {
```
How can there be candidates with null hash?

ghubstan replied:
There can be a null hash each time a user clicks on the view, and calling `OfferBook#fillOfferBookListItems()`.  That method loads the view from `offerBookService.getOffers()`, and the returned list elements are mapped to new `OfferBookListItem(offer)`s with a null hash code.

We discussed this on keybase, when you mentioned the idea of adding a `transient var hash` to OfferPayload.
At the time, I also mentioned that without the cached OfferPayload hash,  OfferBook line 208 might be 
        `.map(o -> new OfferBookListItem(o, get32ByteHashAsByteArray(o.getOfferPayload()))`

There is a comment about null hashes in `OfferBookListItem`:  _A null value indicates the item's payload hash has not been set by onAdded or onRemoved since the most recent OfferBook view refresh._

sqrrm replied:
Does this mean an item with null hash will be removed if only the id matches? Seems like a way to get around the hash checks.

ghubstan replied:
It is a way to get around the hash check if the list contains an item loaded by `OfferBook#fillOfferBookListItems()` (where items are created with null hashes).

If we do create a transient OfferPayload hash field, fillOfferBookListItems() can load the view with non-null hashes, and this null hash check can be removed.

ghubstan committed:
Do not filter on paylaod hash when deleting duplicate list items

Any and all view list items with a matching offerId should be
removed from view just before adding a new list item.

sqrrm commented on desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:
```diff
@@ -83,44 +85,112 @@ public void onAdded(Offer offer) {
+        List<OfferBookListItem> duplicateItems = offerBookListItems.stream()
+                .filter(item -> item.getOffer().getId().equals(offerId))
+                .collect(Collectors.toList());
+        if (duplicateItems.size() > 0) {
```
This check is not needed since `forEach` won't do anything if there are no items in the list.

ghubstan replied:
Removed list.size check in commit 4889da6bba6ad155436a2dede5182643f706ac6d.

sqrrm commented on desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java:
```diff
@@ -83,44 +85,112 @@ public void onAdded(Offer offer) {
+                            offerId,
+                            oldOfferItem.getHashOfPayload() == null ? "null" : oldOfferItem.getHashOfPayload().getHex(),
+                            oldOfferItem.getOffer().getId());
+                }
```
There are still a lot of debug logs. Do you want to keep them for the first release with these changes? If so, I think it would be good to add a TODO to remove them later as I don't think they're that valuable in the long run.

ghubstan replied:
>> RE  _a lot of debug logs_

TODO comments are added in commit 84036bd86212d0eab8bd48a1114985c5b87b14e1.

ghubstan committed:
Remove redundant list.size check

ghubstan committed:
Add missing detail to comment

ghubstan committed:
Add TODOs (delete debug statement)

The new debug log statements included in this PR help trace add/remove
list item actions if problems are seen in the UI's OfferBook, after the
API 'editoffer' method is released.

They can and should be removed in a future PR if the released API feature
proves it did not introduce bugs into the UI.

sqrrm reviewed (approved):
ACK
